core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -386,7 +387,7 @@ public final void runInContext() {
             }
           } catch (Throwable t) {
             stream.cancel(Status.CANCELLED.withCause(t).withDescription("Failed to read message."));
-            return;
+            Throwables.propagateIfPossible(t);
```

Comment:
> AFAICT, as it stands right now, an NPE in an onNext callback is sent on the stream but never noted locally

It's supposed to be the reverse. Only a cancellation is sent to the server and the exception is supposed to be delivered to the client's StreamObserver.onError(). I think what may be happening is since the call is already completed as far as the transport is concerned, the status is squelched. That does seem like a problem, but will be a bit invasive to fix. I'm not sure what solution we'll want short-term.

Note that we've tried to make sure exceptions get handled properly. In the onNext() case we should deliver the exception back to the application. For a case where delivery to the application is not possible, like [onClose (which turns into StreamObserver.onError())](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L421), we do end up propagating the exception further up the call stack to be handled by the executor/UncaughtExceptionHandler.

> In my case, it's the RPC result delivery onNext so the server side doesn't see it? It certainly doesn't care if the client throws and exception, or at least why.

Yep, the server shouldn't care. That's part of the current design.